server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/SortOperator.java

```diff
@@ -67,75 +108,253 @@ public ListenableFuture<?> isBlocked() {
+    ColumnBuilder[] valueColumnBuilders = tsBlockBuilder.getValueColumnBuilders();
+    while (!mergeSortHeap.isEmpty()) {
+
+      MergeSortKey mergeSortKey = mergeSortHeap.poll();
+      TsBlock targetBlock = mergeSortKey.tsBlock;
+      timeBuilder.writeLong(targetBlock.getTimeByIndex(mergeSortKey.rowIndex));
+      for (int i = 0; i < valueColumnBuilders.length; i++) {
+        if (targetBlock.getColumn(i).isNull(mergeSortKey.rowIndex)) {
+          valueColumnBuilders[i].appendNull();
+          continue;
+        }
+        valueColumnBuilders[i].write(targetBlock.getColumn(i), mergeSortKey.rowIndex);
+      }
+      tsBlockBuilder.declarePosition();
+
+      int readerIndex = mergeSortKey.tsBlockIndex;
+      mergeSortKey = readNextMergeSortKey(readerIndex);
+      if (mergeSortKey != null) {
+        mergeSortHeap.push(mergeSortKey);
+      } else {
+        noMoreData[readerIndex] = true;
+        sortBufferManager.releaseOneSortBranch();
+      }
+
+      // break if time is out or tsBlockBuilder is full or sortBuffer is not enough
+      if (System.nanoTime() - startTime > maxRuntime || tsBlockBuilder.isFull()) {
+        break;
+      }
+    }
     return tsBlockBuilder.build();
   }
 
+  private MergeSortKey readNextMergeSortKey(int readerIndex) throws IoTDBException {
+    SortReader sortReader = sortReaders.get(readerIndex);
+    if (sortReader.hasNext()) {
+      MergeSortKey mergeSortKey = sortReader.next();
+      mergeSortKey.tsBlockIndex = readerIndex;
+      return mergeSortKey;
+    }
+    return null;
+  }
+
+  private boolean hasMoreData() {
+    if (noMoreData == null) return true;
+    for (boolean noMore : noMoreData) {
+      if (!noMore) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public void clear() {
+    if (!diskSpiller.hasSpilledData()) return;
+    try {
+      diskSpiller.clear();
+      Path path = Paths.get(filePrePath);
+      Files.deleteIfExists(path);
+      if (sortReaders != null) {
+        for (SortReader sortReader : sortReaders) {
+          sortReader.close();
+        }
+      }
+    } catch (Exception e) {
+      logger.error("delete spilled file error: {}", e.getMessage());
+    }
+  }
+
   @Override
   public boolean hasNext() throws Exception {
-    return inputOperator.hasNextWithTimer() || cachedData != null;
+    return inputOperator.hasNextWithTimer()
+        || (!diskSpiller.hasSpilledData() && curRow != cachedData.size())
+        || (diskSpiller.hasSpilledData() && hasMoreData());
   }
 
   @Override
   public void close() throws Exception {
+    cachedData = null;
+    clear();
     inputOperator.close();
   }
 
   @Override
   public boolean isFinished() throws Exception {
-    return cachedData == null;
+    return inputOperator.isFinished()
+        && (cachedData == null || cachedData.size() == curRow)
+        && ((diskSpiller.hasSpilledData() && !hasMoreData()) || !diskSpiller.hasSpilledData());
```

Comment:
```suggestion
        && (!diskSpiller.hasSpilledData() || !hasMoreData());
```

Reply:
done. The method was changed to `return inputOperator.isFinished()&&!hasNext()`